### ISSUE TITLE
sae_manifold_fit: euclidean multi-atom (K>1) hangs indefinitely on real D=5120 LLM activations

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#1094]
## Observed behaviour

When fitting `sae_manifold_fit` with `atom_topology=\"euclidean\"` and `K≥2` on PCA-whitened OLMo-3-32B L25 residual-stream activations (N=635, D=5120→PCA-10 or PCA-32), the outer BFGS loop stalls after printing the rank-deficiency audit lines and never makes a single outer iteration.  The process-monitor shows 196 threads at ~380 MB RSS with `active_threads=0` — the inner Arrow-Schur solve is deadlocked or spinning without convergence.

Repro (on MSI, after building gamfit):
```
python3 examples/olmo_sae_atlas.py \
    --data_dir /projects/standard/REDACTED/REDACTED/olmo_data \
    --out_dir  /projects/standard/REDACTED/REDACTED/olmo_data/plots \
    --pca_dim 10 --n_atoms 8 --n_iter 30
```

Log excerpt:
```
[OUTER] SAE manifold: n_params=18, gradient=Analytic, hessian=Unavailable -> solver=Bfgs
[SAE-AUDIT] decoder atom 'atom_0' weighted design is rank-deficient (rank=3/6, ...)
... (repeats every 1-2 s indefinitely, no BFGS step ever logged)
[process-monitor] elapsed=2m00s rss=393.5MiB active_threads=0
```

## Root cause hypothesis

The team-lead notes this is the same H_tt non-PD continuation spine issue as #1051 (euclidean + multi-atom = Arrow-Schur block is indefinite at init on real activation geometry).  The inner Newton step for the H_tt block fails to certify positive-definiteness and the solver is looping without accepting any step.

## Workaround applied

Dodge-config: `K=1, atom_topology=\"circle\", --pca_dim 32`.  Single-atom circle fits converge (confirmed running).  Multi-atom euclidean does not.

## Why this matters

This blocks all multi-atom manifold fits on real LLM data.  The SAE dictionary learner's primary use case is K≥4 atoms to decompose different concept clusters.  K=1 is a proof-of-concept only.

## Steps to confirm fix

`sae_manifold_fit(X=pca_olmo_l25, K=4, atom_topology=\"euclidean\", n_iter=30)` should complete within a few minutes on N=635, D=32.

**Comment by SauersML** [comment]
Fixed in 8ab4955b9: when K>1 euclidean is at KKT stationarity and H_tt is non-PD (due to softmax logit-coord GN cross-terms making the logit Schur complement negative at the true inner optimum), fall back to a lightly-damped Direct factor with ridge_ext_coord. The consistent log-det bias across all rho evaluations does not shift the optimal rho.

**Comment by SauersML** [comment]
Reopening. The first real full-suite CI run (post vacuous-shard fix) shows this issue's pinned test is honest-RED: sae_manifold_euclidean_k2_terminates (#1094) / sae_manifold_small_n_circle_seed_accept (#1095). Diagnostic review found the closing commit a266a4e27 delivered only the TERMINATION half (anti-hang guard — real, works: bridges.rs:1002 fires consecutive_probe_refusals→Fatal, killing the 97-min hang) but NOT the 'converge to R²>0.9 / accept seed' half its message claimed (the diff is no-op struct-field plumbing). The remaining failure is the indefinite multi-atom Arrow-Schur recovery / gauge — genuinely research-grade, tracked under #2134/#2111/#2087. Re-scoping here so the red test isn't falsely marked resolved.

**Comment by SauersML** [comment]
## Verified on current `main` — the HANG is dead; a residual **outer-criterion refusal** keeps this red

Drove the production path (`OuterProblem::run` around `SaeManifoldOuterObjective`) on a cold K=2 euclidean fit via the dedicated regression pin `sae_manifold_euclidean_k2_fit_terminates` (`tests/sae/sae/sae_manifold_euclidean_k2_terminates.rs`). Rust CI run **28903085679** (sha `c22d838fb`, ancestor of current HEAD; the only `main` change to `construction.rs` since is an unrelated HT-weight installer, so this is current):

```
[#1094] euclidean K=2 fit: final_value=1.000000e12 recon_R2=0.989227 elapsed=7.2s
panicked at ...:188: euclidean K=2 fit terminated at the infeasible sentinel (final_value=1.000000e12)
```

### What is fixed (the original mechanism)
- **No hang.** The fit **terminates in 7.2 s** (ceiling 150 s; the issue's 240 s deadlock is gone). The inner Arrow–Schur solve no longer spins with `active_threads=0`.
- **The two lines are recovered:** reconstruction **R² = 0.989**.

The inner solve is now provably bounded: every loop level has a hard cap, and the #1094-specific escape is the `consecutive_stall_factor_fail` early-`Err` at `crates/gam-sae/src/manifold/construction.rs:5741-5761` (its comment names this issue: *"multi-atom euclidean with near-zero initial latent coords, #1094 … Without this early return the loop kept burning the extended progress budget indefinitely"*), plus the spectral unit-stiffness deflation of indefinite `H_tt` at `crates/gam-solve/src/arrow_schur/factorization.rs:446`. This is **not** the self-concordant decoder-barrier from 4b198bc72/58df43cfc (that stiffens the β/`H_bb` co-collapse wall, not the t-block).

### What remains (why the pin is still red)
The fit converges to an essentially correct reconstruction (R²=0.99) but the **outer REML criterion returns the infeasible sentinel `1e12`**, not a finite Laplace evidence. Mechanism: at the rank-deficient K=2 fixed point the *undamped* log-evidence factorization cannot certify PD, so `converge_inner_for_undamped_logdet` returns the stall-factor-fail `Err` at **every** ρ probe → `is_recoverable_value_probe_refusal` maps each to `+INF` → the outer BFGS never ranks a finite criterion and reports the sentinel. In other words the very early-`Err` that killed the hang now refuses to score a point that is de-facto converged.

This is the **indefinite log-evidence-block / PD-floor class** (the undamped `½log|H|` refuses on rank-deficient geometry), the same root tracked by #1795 (PD-floor) / #1038 (rank-1 capacitance majorizer) / #2087 (logdet-block desync) — **not** the wide-p cubic outer-REML cost of #2080 (this repro is PCA-reduced: n_params=18, D=10/32, so #2080's width wall does not apply). The correct resolution is to let the outer criterion rank the **PD-floored/deflated** evidence at the rank-deficient fixed point rather than refuse it outright.

**Keeping OPEN** — retargeted from "hangs indefinitely" to "outer criterion returns the infeasible sentinel despite R²=0.99 recovery (rank-deficient K=2 log-evidence refusal)". The hang is closed; the pin fails on the `final_value < 1e11` assertion, not the wall-clock ceiling.

**Comment by SauersML** [comment]
## Fix landed (commit dbdb20d27) — rank the finite deflated evidence at the persistent stall fixed point

### Precise mechanism (confirmed by tracing, not the "factor fails" reading)
The residual refusal is in `converge_inner_for_undamped_logdet` (`crates/gam-sae/src/manifold/construction.rs`). At the rank-deficient K=2 optimum the inner solve reaches R²≈0.99 but the KKT gradient never falls below tolerance — it parks permanently in the weakly-identified decoder/gauge directions the near-singular β-Schur can't resolve. So the clean KKT-stationary exit (`gradient_stationary`, ~5547) is never taken; the loop refines until the **objective** stalls and enters the stall block (~5795).

There, the undamped evidence factor at the stationary iterate **does succeed** with a finite log-det — the per-row `H_tt` spectral deflation (`factor_spectral_deflated_evidence_row`) and the reduced-Schur PD-floor/unit-deflation (`factor_dense_reduced_schur`) guarantee a PD factor on finite geometry. But neither stationarity certificate fires: the raw/quotient gradient is above tol, and the affine-invariant Newton-decrement `½λ²/scale` is above `STALL_REL_TOL` (the near-singular Schur over-predicts a decrease the line search can't realize in the flat direction). So it fell through to `consecutive_stall_factor_fail` and, after 3 rounds, returned the "did not converge" `Err` → `is_recoverable_value_probe_refusal` → +∞ at every ρ → outer BFGS reports the `1e12` sentinel.

So this is *not* "the undamped factor refuses PD" — the factor produces a **finite** Laplace evidence; the driver was discarding it because no stationarity certificate could accept a rank-deficient fixed point whose gradient can't be driven to zero.

### Fix
The penalised objective holding at its numerical floor for `STALL_MIN_ROUNDS` consecutive refine rounds *is* the #1051 inner-convergence witness. When the evidence factor yields a finite log-det at such a round, stash that cache; once the stall persists for the full streak, rank it instead of refusing. The unit-stiffness/PD-floor conditioning of the rank-deficient directions is a bias consistent across every ρ, so it does not move the outer optimum (per 8ab4955b9's note), whereas refusing poisons the whole fit to the sentinel. The hard refusal is retained for the genuinely-broken case where the factor fails at every stall round (no finite evidence to rank); the stash is cleared whenever the stall streak breaks so only a cache from an unbroken frozen-iterate run is ever ranked. This does not touch the two existing certificates, so K=1 circle (#2226) and co-collapse (#2089) accepts are unchanged — the new accept can only convert a *current* sentinel-refusal into a finite ranking, never alter an already-finite path.

The pin `sae_manifold_euclidean_k2_fit_terminates` asserts `final_value < 1e11`; this fix makes the outer criterion finite at the converged K=2 optimum.

### Verification still owed
Uncompiled (8GB shared laptop; agent does not build). Needs `./build.sh` + the regression pin + full SAE suite on CI to confirm `final_value < 1e11` and no regression in the stall/refusal-dependent tests (`tests_schur_seed_refusal_1782`, `tests_cocollapse_reseed_2089`, `tests_outer_reml_probe_budget_2080`, #2226 K=1 circle). Keeping OPEN pending that.

**Comment by SauersML** [comment]
### Test pin extended (commit 205dbb996)

`tests/sae/sae/sae_manifold_euclidean_k2_terminates.rs` previously asserted only the optimizer's reported `final_value < 1e11` and `R² > 0.9`. Extended it to re-evaluate the outer criterion **directly** through the public `SaeManifoldTerm::reml_criterion` path at the fit's own converged ρ (`into_fitted().rho`), and assert that value is also finite and below the infeasible sentinel:

```
assert!(recheck_cost.is_finite() && recheck_cost < 1.0e11,
  "euclidean K=2 outer criterion re-evaluated to the infeasible sentinel at the converged \
   optimum (recheck_criterion={recheck_cost:.6e}) despite R²={r2:.6} ...");
```

This pins the criterion PATH itself (not just the optimizer's bookkeeping value) returning a finite Laplace evidence score for a feasible K=2 fit that reconstructs the two planted lines — the exact "good fit reports infeasible" failure mode. Fix in dbdb20d27, test in 205dbb996; both local, unpushed, pending compile+CI.

**Comment by SauersML** [comment]
Closing on HEAD-verified evidence (audit vs current main + the full 10/10-green workspace sweep at 5a422a09a, run 28987346771).

Two-part fix, both on HEAD: (1) the K>1 indefinite-`H_tt`-at-stationarity handling (a266a4e27) and (2) the persistent-objective-stall accept (dbdb20d27): in `converge_inner_for_undamped_logdet` the stall path now, after `SAE_MANIFOLD_INNER_OBJECTIVE_STALL_MIN_ROUNDS` consecutive stalled rounds, returns the finite deflated Laplace evidence (`stalled_finite_cache`) instead of erroring to the 1e12 infeasible sentinel; the stall streak resets whenever a round genuinely descends, so only a frozen iterate is accepted. The acceptance-reachability siblings (37e4d2eae, 36bec0e29) force-install per-row spectral deflation before every ridge-0 acceptance factorization.

Validating test, live and green in the full sweep: `sae_manifold_euclidean_k2_fit_terminates` (tests/sae/sae/sae_manifold_euclidean_k2_terminates.rs) — fits K=2 planted euclidean lines, asserts wall-clock bounded, R² > 0.9, and that re-evaluating `reml_criterion` at the converged ρ is finite and < 1e11 (not the sentinel), pinning the criterion path per 205dbb996.

(The circle-topology d_atom=2 quality follow-up — fit completes but under-recovers — is a distinct defect tracked in #2228, under active work.)
